### PR TITLE
change(Balsamic): Analysis on normal samples

### PR DESCRIPTION
**Pull request by ivadym**
### Changed
- Balsamic analysis taking normal samples as input.
- Delivery report minor formatting changes.

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [x] `cg workflow balsamic start <case_id>` (tumor-only analysis)
Screenshot 2023-02-03 at 09 06 35

- [x] `cg workflow balsamic start <case_id>` (tumor-normal analysis)
Screenshot 2023-02-03 at 09 08 12

- [x] `cg workflow balsamic start <case_id>` (2 normal sample)
Screenshot 2023-02-03 at 09 34 16

- [x] `cg workflow balsamic start <case_id>` (1 normal sample)
Screenshot 2023-02-03 at 09 09 13

- [x] `cg workflow balsamic start <case_id>` (1 normal sample, `--force-normal`)
Screenshot 2023-02-03 at 09 29 48

- Delivery report
Screenshot 2023-02-06 at 11 54 27

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/meta/workflow/balsamic.py [hunk 1]
```diff
@@ -316,27 +296,55 @@ def get_verified_gender(sample_data: dict) -> Union[Gender.FEMALE, Gender.MALE]:
...
-    def get_verified_normal_path(sample_data: dict) -> Optional[str]:
-        """Takes a dict with samples and attributes, and retrieves the path
-        of normal sample.
-        If the number of paths is exactly 1, the path is returned.
-        If there are no paths, then the sample is not paired, and None is returned.
-        Raises BalsamicStartError:
-            When there is more than one normal sample.
-        """
-        normal_paths = [
+    def get_verified_sample_path(sample_data: dict, sample_type: SampleType) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (Vince-janv) [reply cg/meta/workflow/balsamic.py:334]: I've looked a bit more and maybe we can define a simple dataclass (https://realpython.com/python-data-classes/) for a sample_config? That combined with making the methods more modular would really improve readability

File: cg/constants/constants.py [hunk 2]
```diff
@@ -99,6 +99,11 @@ class GenomeVersion(StrEnum):
     canfam3: str = "canfam3"
 
 
+class SampleType(StrEnum):
+    TUMOR: str = "tumor"
+    NORMAL: str = "normal"
```
Comment (Vince-janv) [inline cg/constants/constants.py:104]: 🌟

**Review by Vince-janv — COMMENTED** [review]
I not well versed in the balsamic workflow code, but I found it quite hard to follow. Maybe the config part can be simplified? This might be a bit out of scope though since you only touched a small bit of it 😅

File: cg/meta/upload/scout/balsamic_config_builder.py [hunk 3]
```diff
@@ -36,10 +38,10 @@ def include_sample_files(self, config_sample: ScoutBalsamicIndividual) -> None:
 
         sample_id: str = config_sample.sample_id
         if config_sample.alignment_path:
-            if "tumor" in config_sample.alignment_path:
-                sample_id = "tumor"
-            elif "normal" in config_sample.alignment_path:
-                sample_id = "normal"
+            if SampleType.TUMOR.value in config_sample.alignment_path:
+                sample_id = SampleType.TUMOR.value
```
Comment (fevac) [inline cg/meta/upload/scout/balsamic_config_builder.py:42]: For consistency keep it always either as "SampleType.TUMOR/NORMAL.value" or "SampleType.TUMOR/NORMAL", sometimes is specified in one way and sometimes in the other. I'd go for the second one.

File: cg/meta/workflow/balsamic.py [hunk 4]
```diff
@@ -177,11 +177,11 @@ def get_gender(sample_obj: models.Sample) -> str:
...
 
     @staticmethod
-    def get_sample_type(sample_obj: models.Sample) -> str:
-        """Returns tissue type of a sample"""
+    def get_sample_type(sample_obj: models.Sample) -> SampleType:
+        """Return the tissue type of the sample."""
         if sample_obj.is_tumour:
-            return "tumor"
-        return "normal"
+            return SampleType.TUMOR
```
Comment (fevac) [inline cg/meta/workflow/balsamic.py:183]: Same as before: "SampleType.TUMOR/NORMAL.value" or "SampleType.TUMOR/NORMAL"

File: cg/store/status_sample_filters.py [hunk 5]
```diff
@@ -1,21 +1,43 @@
...
+) -> Query:
     """Apply filtering functions to the sample queries and return filtered results."""
     filter_map = {
+        "samples_with_case_id": get_samples_with_case_id,
+        "samples_with_type": get_samples_with_type,
         "samples_uploaded_to_loqusdb": get_samples_with_loqusdb_id,
         "samples_not_uploaded_to_loqusdb": get_samples_without_loqusdb_id,
     }
-    return filter_map[function](samples=samples)
+    return filter_map[function](samples=samples, case_id=case_id, sample_type=sample_type)
```
Comment (fevac) [inline cg/store/status_sample_filters.py:43]: I'm guessing having `case_id` and `sample_type` as None when not provided doesn't change the filter functionality for other pipelines, right?

File: tests/cli/workflow/balsamic/test_cli_balsamic_config_case.py [hunk 6]
```diff
@@ -22,7 +22,7 @@ def test_without_options(cli_runner: CliRunner, balsamic_context: CGConfig):
 
 
 def test_with_missing_case(cli_runner: CliRunner, balsamic_context: CGConfig, caplog):
```
Comment (fevac) [inline tests/cli/workflow/balsamic/test_cli_balsamic_config_case.py:24]: type hint `caplog: LogCaptureFixture` and the same for the functions below

File: cg/meta/upload/scout/balsamic_config_builder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivadym) [reply cg/meta/upload/scout/balsamic_config_builder.py:42]: No, it's not! Good call 👍

File: cg/store/api/findbusinessdata.py [hunk 7]
```diff
@@ -384,8 +385,19 @@ def samples_by_ids(self, **identifiers) -> Query:
     def get_sample_by_name(self, name: str) -> Sample:
         return self.Sample.query.filter(Sample.name == name).first()
 
+    def get_sample_name_by_type(self, case_id: str, sample_type: SampleType) -> Optional[str]:
+        """Extract sample name given a tissue type."""
```
[the remark this answers is not in the document]
Comment (ivadym) [reply cg/store/api/findbusinessdata.py:389]: 👍

File: cg/store/api/findbusinessdata.py [hunk 8]
```diff
@@ -384,8 +385,19 @@ def samples_by_ids(self, **identifiers) -> Query:
     def get_sample_by_name(self, name: str) -> Sample:
         return self.Sample.query.filter(Sample.name == name).first()
 
+    def get_sample_name_by_type(self, case_id: str, sample_type: SampleType) -> Optional[str]:
+        """Extract sample name given a tissue type."""
+        records: Query = self.Sample.query.join(Family.links, FamilySample.sample)
```
[the remark this answers is not in the document]
Comment (ivadym) [reply cg/store/api/findbusinessdata.py:390]: 👍

File: cg/meta/workflow/balsamic.py [hunk 9]
```diff
@@ -316,27 +296,55 @@ def get_verified_gender(sample_data: dict) -> Union[Gender.FEMALE, Gender.MALE]:
...
+            sample_data=sample_data, sample_type=SampleType.NORMAL
+        )
+
+        if not tumor_path and not normal_path:
+            LOG.error(f"No tumor or normal samples were found for case {case_id}")
+            raise BalsamicStartError
+
+        if normal_path and not tumor_path:
+            LOG.warning(
+                f"Only a normal sample was found for case {case_id}. "
```
Comment (fevac) [inline cg/meta/workflow/balsamic.py:320]: This is good, but in production we do not check all warnings unless there's something weird going on. Maybe it would be nice to have this information in the delivery report. We could specify there that a normal sample has been treated as a tumor and that this is NOT validated in balsamic and at the user own risk. Or something like that. But this could be added later in a different PR.

**Review by fevac — CHANGES_REQUESTED** [review]
Amazing work!! 🌟 🌟  🌟  My main concern is that the "normal as tumor" warning goes unnoticed. I made some comments below

File: cg/meta/workflow/balsamic.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivadym) [reply cg/meta/workflow/balsamic.py:334]: I agree, actually I wanted to change how the Balsamic config case is done (we already talk a bit about it during summer with the PON PR). We are validating plenty of fields, which tell me that we could also use pydantic models and simplify this quite a lot.

That said, I would not to address it in this PR. `get_verified_sample_path` for example will be removed in the next one when I remove concatenation.

[the remark this answers is not in the document]
Comment (ivadym) [reply cg/meta/workflow/balsamic.py:334]: If you agree, I'll open an issue and try to address it in the next few weeks, it's time to refactor this taking advantage of the fact that we're going to remove the balsamic concatenation from cg

File: cg/meta/upload/scout/balsamic_config_builder.py (hunk 3 above)
Comment (ivadym) [reply cg/meta/upload/scout/balsamic_config_builder.py:42]: There is a reason behind it:
Screenshot 2023-01-31 at 11 02 07

We already had problems before in production with this, so I'm trying to keep the assignments with `.value` and  leave them as enum types in conditional statements.

File: cg/store/status_sample_filters.py (hunk 5 above)
Comment (ivadym) [reply cg/store/status_sample_filters.py:43]: Good point, but it doesn't in this case as these arguments are optional and they take None as default:
```
    case_id: Optional[str] = None,
    sample_type: Optional[SampleType] = None,
```

File: tests/cli/workflow/balsamic/test_cli_balsamic_config_case.py (hunk 6 above)
Comment (ivadym) [reply tests/cli/workflow/balsamic/test_cli_balsamic_config_case.py:24]: Changed ☺️

File: cg/meta/workflow/balsamic.py (hunk 9 above)
Comment (ivadym) [reply cg/meta/workflow/balsamic.py:320]: I would say that this PR is a hot fix and should be reverted as soon as we figure out why the customers need to run Balsamic analyses on normal samples.

Now the question is, should it even be picked by the automation or should we only allow these cases to be started manually? What do you think?

Comment (fevac) [reply cg/meta/workflow/balsamic.py:320]: I don't see why we would revert this, this might come back again in the future. But I agree that we would need to take care of the automation. I would say we do not want this to be picked by the automation UNLESS there are many of these types of cases (which would result in a lot of manual work for people in prod). But possibly @karlnyr knows better

File: cg/meta/upload/scout/balsamic_config_builder.py (hunk 3 above)
Comment (fevac) [reply cg/meta/upload/scout/balsamic_config_builder.py:42]: 😮 !! Good to know!

File: cg/meta/upload/scout/balsamic_config_builder.py [hunk 10]
```diff
@@ -50,12 +52,12 @@ def build_config_sample(self, db_sample: models.FamilySample) -> ScoutBalsamicIn
         config_sample = ScoutBalsamicIndividual()
 
         self.add_common_sample_info(config_sample=config_sample, db_sample=db_sample)
-        if BalsamicAnalysisAPI.get_sample_type(db_sample.sample) == "tumor":
-            config_sample.phenotype = "affected"
-            config_sample.sample_id = "TUMOR"
+        if BalsamicAnalysisAPI.get_sample_type(db_sample.sample) == SampleType.TUMOR:
```
[the remark this answers is not in the document]
Comment (ivadym) [reply cg/meta/upload/scout/balsamic_config_builder.py]: 👍 Changed it for all over these methods

File: cg/meta/workflow/balsamic.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vince-janv) [reply cg/meta/workflow/balsamic.py:334]: That sounds good! ⭐

File: cg/meta/upload/scout/balsamic_config_builder.py [hunk 11]
```diff
@@ -32,32 +34,32 @@ def include_case_files(self):
...
+            elif SampleType.NORMAL in config_sample.alignment_path:
+                sample_id = SampleType.NORMAL.value
 
         config_sample.vcf2cytosure = self.fetch_sample_file(
             hk_tags=self.sample_tags.vcf2cytosure, sample_id=sample_id
         )
 
-    def build_config_sample(self, db_sample: models.FamilySample) -> ScoutBalsamicIndividual:
-        """Build a sample with balsamic specific information"""
+    def build_config_sample(self, family_sample: models.FamilySample) -> ScoutBalsamicIndividual:
```
Comment (henrikstranneheim) [inline cg/meta/upload/scout/balsamic_config_builder.py:50]: Can we use `case_sample`instead. family does not really make sense for a cancer analysis
```suggestion
    def build_config_sample(self, case_sample: FamilySample) -> ScoutBalsamicIndividual:
```

File: cg/meta/upload/scout/mip_config_builder.py [hunk 12]
```diff
@@ -77,13 +77,13 @@ def include_pedigree_picture(self) -> None:
         else:
             LOG.info("family of 1 sample - skip pedigree graph")
 
-    def build_config_sample(self, db_sample: models.FamilySample) -> ScoutMipIndividual:
+    def build_config_sample(self, family_sample: models.FamilySample) -> ScoutMipIndividual:
```
Comment (henrikstranneheim) [inline cg/meta/upload/scout/mip_config_builder.py:80]:
```suggestion
    def build_config_sample(self, case_sample: FamilySample) -> ScoutMipIndividual:
```

File: cg/meta/upload/scout/mip_config_builder.py [hunk 13]
```diff
@@ -77,13 +77,13 @@ def include_pedigree_picture(self) -> None:
...
         """Build a sample with mip specific information"""
 
         config_sample = ScoutMipIndividual()
-        self.add_common_sample_info(config_sample=config_sample, db_sample=db_sample)
-        config_sample.father = db_sample.father.internal_id if db_sample.father else "0"
-        config_sample.mother = db_sample.mother.internal_id if db_sample.mother else "0"
+        self.add_common_sample_info(config_sample=config_sample, family_sample=family_sample)
+        config_sample.father = family_sample.father.internal_id if family_sample.father else "0"
+        config_sample.mother = family_sample.mother.internal_id if family_sample.mother else "0"
```
Comment (henrikstranneheim) [inline cg/meta/upload/scout/mip_config_builder.py:86]:
```suggestion
        config_sample.father = family_sample.father.internal_id if family_sample.father else RelationshipStatus.HAS_NO_PARENT
        config_sample.mother = family_sample.mother.internal_id if family_sample.mother else RelationshipStatus.HAS_NO_PARENT
```

File: cg/store/api/findbusinessdata.py [hunk 14]
```diff
@@ -384,8 +385,18 @@ def samples_by_ids(self, **identifiers) -> Query:
     def get_sample_by_name(self, name: str) -> Sample:
         return self.Sample.query.filter(Sample.name == name).first()
 
+    def get_samples_by_type(self, case_id: str, sample_type: SampleType) -> Optional[List[Sample]]:
+        """Extract samples given a tissue type."""
+        samples: Query = self.Sample.query.join(Family.links, FamilySample.sample)
```
Comment (henrikstranneheim) [inline cg/store/api/findbusinessdata.py:390]: If you have time - it would be great to separate the query statement into a separate function in status.py. Much like
```
    def _get_sample_query(self) -> Query:
        """Return sample query."""
        return self.query(Sample)
```
But of course reflect that this sample query contains `Family` and `FamilySample` as well.

**Review by henrikstranneheim — APPROVED** [review]
Great work @ivadym! Fantastic clean-up as well. 💯.

I agree with that we should not add this to the automation first thing. Let's wait and see

File: cg/store/api/findbusinessdata.py (hunk 14 above)
Comment (ivadym) [reply cg/store/api/findbusinessdata.py]: Good idea 👍

Comment (ivadym) [reply cg/store/api/findbusinessdata.py]: It's a good idea @henrikstranneheim, but I'm not sure how to proceed with it... 

We already have a newly added `_get_case_sample_query` in this file, which I suppose has been included here because of how we have structured `Store` and `CoreHandler`, i.e. not allowing methods to be used between these StatusHandler, FindBusinessDataHandler, etc. classes.

To bypass it, we could create some inheritances between the handlers, having StatusHandler as a parent class for example, but I'm not sure how this would affect the Store class and if it would make sense to have both implementations.

File: cg/meta/workflow/balsamic.py (hunk 9 above)
Comment (ivadym) [reply cg/meta/workflow/balsamic.py]: I've added a `--force-normal` flag, which will prevent these analyses from starting automatically. We can also easily remove it if we start getting too many of these cases.

**Review by fevac — APPROVED** [review]

File: cg/store/api/findbusinessdata.py (hunk 14 above)
Comment (henrikstranneheim) [reply cg/store/api/findbusinessdata.py]: Sorry my bad, I meant that you could add the new method within the same module i.e in `findbusinessdata.py`

**Comment by ivadym** [comment]
Screenshot 2023-02-06 at 12 05 06
